docs/config.md
# Virtlet configuration

Virtlet has a flexible configuration mechanism that makes it possible
to specify per-node configuration using CRDs.

## Using per-node configuration

In order to use per-node configuration, you need to create Virtlet
CRD definitions before you deploy Virtlet:

```bash
virtletctl gen --crd | kubectl apply -f -
```

After that, you can add one or more Virtlet configuration mappings:
```bash
kubectl apply -f - <<EOF
---
apiVersion: "virtlet.k8s/v1"
kind: VirtletConfigMapping
metadata:
  name: test
  namespace: kube-system
spec:
  nodeName: kube-node-2
  priority: 1
  config:
    disableKVM: true
    logLevel: 5
    rawDevices: sda*,loop*
---
apiVersion: "virtlet.k8s/v1"
kind: VirtletConfigMapping
metadata:
  name: test-labels
  namespace: kube-system
spec:
  nodeSelector:
    foo: bar
  priority: 10
  config:
    logLevel: 3
EOF
```

You can delete Virtlet pods to have them restarted and pick up the
changes after you add, remove or modify the configuration mappings.

All the config mappings must reside in `kube-system` namespace.  Each
mapping can specify `nodeSelector` or `nodeName` to target a subset of
the nodes. If neither `nodeSelector` nor `nodeName` is specified, the
mapping will target all the nodes. If the several mappings apply to
the node, their order is determined by `priority` value, with mappings
with higher value of `priority` taking precendence. `config` specifies
the list of configuration fields (see the full table of config fields
in the next section of this document).

In the example above, for the node named `kube-node-2`, KVM is
disabled, log level is set to 5 and the raw device list is set to
`sda*,loop*`.  For nodes with `foo=bar` label, the log level is
overridden and set to 3.  If `kube-node-2` node has `foo=bar` label,
the log level will be set to 3 there as the label-based mapping has
higher priority.

## Virtlet configuration summary

The table below lists all the Virtlet configuration fields, with their
corresponding command line flags and environment variables that are
handled by Virtlet binary. Note that you may only need these
environment variables and configuration flags if you're not using the
standard Virtlet deployment YAML as generated by `virtletctl gen`.

<!--
    The part of this file between 'begin' / 'end' comments is updated
    automatically by build/cmd.sh update-docs
-->
<!-- begin -->
| Description | Config field | Default value | Type | Command line flag / Env |
| --- | --- | --- | --- | --- |
| Path to fd server socket | `fdServerSocketPath` | `/var/lib/virtlet/tapfdserver.sock` | string | `--fd-server-socket-path` / `VIRTLET_FD_SERVER_SOCKET_PATH` |
| Path to the virtlet database | `databasePath` | `/var/lib/virtlet/virtlet.db` | string | `--database-path` / `VIRTLET_DATABASE_PATH` |
| Image download protocol. Can be https or http | `downloadProtocol` | `https` | string | `--image-download-protocol` / `VIRTLET_DOWNLOAD_PROTOCOL` |
| Image directory | `imageDir` | `/var/lib/virtlet/images` | string | `--image-dir` / `VIRTLET_IMAGE_DIR` |
| Image name translation configs directory | `imageTranslationConfigsDir` | `/etc/virtlet/images` | string | `--image-translation-configs-dir` / `VIRTLET_IMAGE_TRANSLATIONS_DIR` |
| Libvirt connection URI | `libvirtURI` | `qemu:///system` | string | `--libvirt-uri` / `VIRTLET_LIBVIRT_URI` |
| Comma separated list of raw device glob patterns which VMs can access (without '/dev/' prefix) | `rawDevices` | `loop*` | string | `--raw-devices` / `VIRTLET_RAW_DEVICES` |
| The path to UNIX domain socket for CRI service to listen on | `criSocketPath` | `/run/virtlet.sock` | string | `--listen` / `VIRTLET_CRI_SOCKET_PATH` |
| Display logging and the streamer | `disableLogging` | `false` | boolean | `--disable-logging` / `VIRTLET_DISABLE_LOGGING` |
| Forcibly disable KVM support | `disableKVM` | `false` | boolean | `--disable-kvm` / `VIRTLET_DISABLE_KVM` |
| Enable SR-IOV support | `enableSriov` | `false` | boolean | `--enable-sriov` / `VIRTLET_SRIOV_SUPPORT` |
| Path to CNI plugin binaries | `cniPluginDir` | `/opt/cni/bin` | string | `--cni-bin-dir` / `VIRTLET_CNI_PLUGIN_DIR` |
| Path to the CNI configuration directory | `cniConfigDir` | `/etc/cni/net.d` | string | `--cni-conf-dir` / `VIRTLET_CNI_CONFIG_DIR` |
| Calico subnet size to use | `calicoSubnetSize` | `24` | integer | `--calico-subnet-size` / `VIRTLET_CALICO_SUBNET` |
| Enable regexp image name translation | `enableRegexpImageTranslation` | `true` | boolean | `--enable-regexp-image-translation` / `IMAGE_REGEXP_TRANSLATION` |
| CPU model to use in libvirt domain definition (libvirt's default value will be used if not set) | `cpuModel` |  | string | `--cpu-model` / `VIRTLET_CPU_MODEL` |
| configurable port to the virtlet server | `streamPort` | `10010` | integer | `--stream-port` / `VIRTLET_STREAM_PORT` |
| Pod's root dir in kubelet | `kubeletRootDir` | `/var/lib/kubelet/pods` | string | `--kubelet-root-dir` / `KUBELET_ROOT_DIR` |
| Log level to use | `logLevel` | `1` | integer | `--v` / `VIRTLET_LOGLEVEL` |
<!-- end -->

Only the following config fields mentioned in this table can be used
with standard Virtlet deployment YAML: `downloadProtocol`,
`rawDevices`, `disableKVM`, `enableSriov`, `calicoSubnetSize`,
`enableRegexpImageTranslation`, `cpuModel` and `logLevel`. Other options may need
adjusting the YAML to change the paths of volume
mounts. `disableLogging` option is intended for debugging purposes
only.


docs/reference/config.md
# Using per-node configuration

In order to use per-node configuration, you need to create Virtlet
CRD definitions before you deploy Virtlet:

```bash
virtletctl gen --crd | kubectl apply -f -
```

After that, you can add one or more Virtlet configuration mappings:
```bash
kubectl apply -f - <<EOF
---
apiVersion: "virtlet.k8s/v1"
kind: VirtletConfigMapping
metadata:
  name: test
  namespace: kube-system
spec:
  nodeName: kube-node-2
  priority: 1
  config:
    disableKVM: true
    logLevel: 5
    rawDevices: sda*,loop*
---
apiVersion: "virtlet.k8s/v1"
kind: VirtletConfigMapping
metadata:
  name: test-labels
  namespace: kube-system
spec:
  nodeSelector:
    foo: bar
  priority: 10
  config:
    logLevel: 3
EOF
```

You can delete Virtlet pods to have them restarted and pick up the
changes after you add, remove or modify the configuration mappings.

All the config mappings must reside in `kube-system` namespace.  Each
mapping can specify `nodeSelector` or `nodeName` to target a subset of
the nodes. If neither `nodeSelector` nor `nodeName` is specified, the
mapping will target all the nodes. If the several mappings apply to
the node, their order is determined by `priority` value, with mappings
with higher value of `priority` taking precendence. `config` specifies
the list of configuration fields (see the full table of config fields
in the next section of this document).

In the example above, for the node named `kube-node-2`, KVM is
disabled, log level is set to 5 and the raw device list is set to
`sda*,loop*`.  For nodes with `foo=bar` label, the log level is
overridden and set to 3.  If `kube-node-2` node has `foo=bar` label,
the log level will be set to 3 there as the label-based mapping has
higher priority.

# Configuration options summary

The table below lists all the Virtlet configuration fields, with their
corresponding command line flags and environment variables that are
handled by Virtlet binary. Note that you may only need these
environment variables and configuration flags if you're not using the
standard Virtlet deployment YAML as generated by `virtletctl gen`.

<!--
    The part of this file between 'begin' / 'end' comments is updated
    automatically by build/cmd.sh update-docs
-->
<!-- begin -->
| Description | Config field | Default value | Type | Command line flag / Env |
| --- | --- | --- | --- | --- |
| Path to fd server socket | `fdServerSocketPath` | `/var/lib/virtlet/tapfdserver.sock` | string | `--fd-server-socket-path` / `VIRTLET_FD_SERVER_SOCKET_PATH` |
| Path to the virtlet database | `databasePath` | `/var/lib/virtlet/virtlet.db` | string | `--database-path` / `VIRTLET_DATABASE_PATH` |
| Image download protocol. Can be https or http | `downloadProtocol` | `https` | string | `--image-download-protocol` / `VIRTLET_DOWNLOAD_PROTOCOL` |
| Image directory | `imageDir` | `/var/lib/virtlet/images` | string | `--image-dir` / `VIRTLET_IMAGE_DIR` |
| Image name translation configs directory | `imageTranslationConfigsDir` | `/etc/virtlet/images` | string | `--image-translation-configs-dir` / `VIRTLET_IMAGE_TRANSLATIONS_DIR` |
| Libvirt connection URI | `libvirtURI` | `qemu:///system` | string | `--libvirt-uri` / `VIRTLET_LIBVIRT_URI` |
| Comma separated list of raw device glob patterns which VMs can access (without '/dev/' prefix) | `rawDevices` | `loop*` | string | `--raw-devices` / `VIRTLET_RAW_DEVICES` |
| The path to UNIX domain socket for CRI service to listen on | `criSocketPath` | `/run/virtlet.sock` | string | `--listen` / `VIRTLET_CRI_SOCKET_PATH` |
| Display logging and the streamer | `disableLogging` | `false` | boolean | `--disable-logging` / `VIRTLET_DISABLE_LOGGING` |
| Forcibly disable KVM support | `disableKVM` | `false` | boolean | `--disable-kvm` / `VIRTLET_DISABLE_KVM` |
| Enable SR-IOV support | `enableSriov` | `false` | boolean | `--enable-sriov` / `VIRTLET_SRIOV_SUPPORT` |
| Path to CNI plugin binaries | `cniPluginDir` | `/opt/cni/bin` | string | `--cni-bin-dir` / `VIRTLET_CNI_PLUGIN_DIR` |
| Path to the CNI configuration directory | `cniConfigDir` | `/etc/cni/net.d` | string | `--cni-conf-dir` / `VIRTLET_CNI_CONFIG_DIR` |
| Calico subnet size to use | `calicoSubnetSize` | `24` | integer | `--calico-subnet-size` / `VIRTLET_CALICO_SUBNET` |
| Enable regexp image name translation | `enableRegexpImageTranslation` | `true` | boolean | `--enable-regexp-image-translation` / `IMAGE_REGEXP_TRANSLATION` |
| CPU model to use in libvirt domain definition (libvirt's default value will be used if not set) | `cpuModel` |  | string | `--cpu-model` / `VIRTLET_CPU_MODEL` |
| configurable port to the virtlet server | `streamPort` | `10010` | integer | `--stream-port` / `VIRTLET_STREAM_PORT` |
| Log level to use | `logLevel` | `1` | integer | `--v` / `VIRTLET_LOGLEVEL` |
<!-- end -->

Only the following config fields mentioned in this table can be used
with standard Virtlet deployment YAML: `downloadProtocol`,
`rawDevices`, `disableKVM`, `enableSriov`, `calicoSubnetSize`,
`enableRegexpImageTranslation`, `cpuModel` and `logLevel`. Other options
may need adjusting the YAML to change the paths of volume
mounts. `disableLogging` option is intended for debugging purposes
only.


pkg/config/config.go
/*
Copyright 2018 Mirantis

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package config

import (
	"fmt"
	"math"
	"sort"

	virtletclient "github.com/Mirantis/virtlet/pkg/client/clientset/versioned"
	flag "github.com/spf13/pflag"
	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/tools/clientcmd"

	virtlet_v1 "github.com/Mirantis/virtlet/pkg/api/virtlet.k8s/v1"
)

const (
	defaultFDServerSocketPath = "/var/lib/virtlet/tapfdserver.sock"
	fdServerSocketPathEnv     = "VIRTLET_FD_SERVER_SOCKET_PATH"

	defaultDatabasePath = "/var/lib/virtlet/virtlet.db"
	databasePathEnv     = "VIRTLET_DATABASE_PATH"

	defaultDownloadProtocol  = "https"
	imageDownloadProtocolEnv = "VIRTLET_DOWNLOAD_PROTOCOL"

	defaultImageDir = "/var/lib/virtlet/images"
	imageDirEnv     = "VIRTLET_IMAGE_DIR"

	defaultImageTranslationConfigsDir = "/etc/virtlet/images"
	imageTranslationsConfigDirEnv     = "VIRTLET_IMAGE_TRANSLATIONS_DIR"

	defaultLibvirtURI = "qemu:///system"
	libvirtURIEnv     = "VIRTLET_LIBVIRT_URI"

	defaultRawDevices = "loop*"
	rawDevicesEnv     = "VIRTLET_RAW_DEVICES"

	defaultCRISocketPath = "/run/virtlet.sock"
	criSocketPathEnv     = "VIRTLET_CRI_SOCKET_PATH"

	disableLoggingEnv = "VIRTLET_DISABLE_LOGGING"
	disableKVMEnv     = "VIRTLET_DISABLE_KVM"
	enableSriovEnv    = "VIRTLET_SRIOV_SUPPORT"

	defaultCNIPluginDir = "/opt/cni/bin"
	cniPluginDirEnv     = "VIRTLET_CNI_PLUGIN_DIR"

	defaultCNIConfigDir = "/etc/cni/net.d"
	cniConfigDirEnv     = "VIRTLET_CNI_CONFIG_DIR"

	defaultCalicoSubnet = 24
	calicoSubnetEnv     = "VIRTLET_CALICO_SUBNET"

	enableRegexpImageTranslationEnv = "IMAGE_REGEXP_TRANSLATION"
	logLevelEnv                     = "VIRTLET_LOGLEVEL"

	defaultCPUModel = ""
	cpuModelEnv     = "VIRTLET_CPU_MODEL"

	defaultStreamPort = 10010
	streamPortEnv     = "VIRTLET_STREAM_PORT"

	kubeletRootDir    = "/var/lib/kubelet/pods"
	kubeletRootDirEnv = "KUBELET_ROOT_DIR"
)

func configFieldSet(c *virtlet_v1.VirtletConfig) *fieldSet {
	var fs fieldSet
	fs.addStringField("fdServerSocketPath", "fd-server-socket-path", "", "Path to fd server socket", fdServerSocketPathEnv, defaultFDServerSocketPath, &c.FDServerSocketPath)
	fs.addStringField("databasePath", "database-path", "", "Path to the virtlet database", databasePathEnv, defaultDatabasePath, &c.DatabasePath)
	fs.addStringFieldWithPattern("downloadProtocol", "image-download-protocol", "", "Image download protocol. Can be https or http", imageDownloadProtocolEnv, defaultDownloadProtocol, "^https?$", &c.DownloadProtocol)
	fs.addStringField("imageDir", "image-dir", "", "Image directory", imageDirEnv, defaultImageDir, &c.ImageDir)
	fs.addStringField("imageTranslationConfigsDir", "image-translation-configs-dir", "", "Image name translation configs directory", imageTranslationsConfigDirEnv, defaultImageTranslationConfigsDir, &c.ImageTranslationConfigsDir)
	// SkipImageTranslation doesn't have corresponding flag or env var as it's only used by tests
	fs.addBoolField("skipImageTranslation", "", "", "", "", false, &c.SkipImageTranslation)
	fs.addStringField("libvirtURI", "libvirt-uri", "", "Libvirt connection URI", libvirtURIEnv, defaultLibvirtURI, &c.LibvirtURI)
	fs.addStringField("rawDevices", "raw-devices", "", "Comma separated list of raw device glob patterns which VMs can access (without '/dev/' prefix)", rawDevicesEnv, defaultRawDevices, &c.RawDevices)
	fs.addStringField("criSocketPath", "listen", "", "The path to UNIX domain socket for CRI service to listen on", criSocketPathEnv, defaultCRISocketPath, &c.CRISocketPath)
	fs.addBoolField("disableLogging", "disable-logging", "", "Display logging and the streamer", disableLoggingEnv, false, &c.DisableLogging)
	fs.addBoolField("disableKVM", "disable-kvm", "", "Forcibly disable KVM support", disableKVMEnv, false, &c.DisableKVM)
	fs.addBoolField("enableSriov", "enable-sriov", "", "Enable SR-IOV support", enableSriovEnv, false, &c.EnableSriov)
	fs.addStringField("cniPluginDir", "cni-bin-dir", "", "Path to CNI plugin binaries", cniPluginDirEnv, defaultCNIPluginDir, &c.CNIPluginDir)
	fs.addStringField("cniConfigDir", "cni-conf-dir", "", "Path to the CNI configuration directory", cniConfigDirEnv, defaultCNIConfigDir, &c.CNIConfigDir)
	fs.addIntField("calicoSubnetSize", "calico-subnet-size", "", "Calico subnet size to use", calicoSubnetEnv, defaultCalicoSubnet, 0, 32, &c.CalicoSubnetSize)
	fs.addBoolField("enableRegexpImageTranslation", "enable-regexp-image-translation", "", "Enable regexp image name translation", enableRegexpImageTranslationEnv, true, &c.EnableRegexpImageTranslation)
	fs.addStringField("cpuModel", "cpu-model", "", "CPU model to use in libvirt domain definition (libvirt's default value will be used if not set)", cpuModelEnv, defaultCPUModel, &c.CPUModel)
	fs.addIntField("streamPort", "stream-port", "", "configurable port to the virtlet server", streamPortEnv, defaultStreamPort, 1, 65535, &c.StreamPort)
	fs.addStringField("kubeletRootDir", "kubelet-root-dir", "", "Pod's root dir in kubelet", kubeletRootDirEnv, kubeletRootDir, &c.KubeletRootDir)
	// this field duplicates glog's --v, so no option for it, which is signified
	// by "+" here (it's only for doc)
	fs.addIntField("logLevel", "+v", "", "Log level to use", logLevelEnv, 6, 0, math.MaxInt32, &c.LogLevel)
	return &fs
}

// GetDefaultConfig returns a VirtletConfig with all values set to default
func GetDefaultConfig() *virtlet_v1.VirtletConfig {
	var c virtlet_v1.VirtletConfig
	configFieldSet(&c).applyDefaults()
	return &c
}

// Override replaces the values in the target config with those
// which are set in the other config.
func Override(target, other *virtlet_v1.VirtletConfig) {
	configFieldSet(target).override(configFieldSet(other))
}

// DumpEnv returns a string with environment variable settings
// corresponding to the VirtletConfig.
func DumpEnv(c *virtlet_v1.VirtletConfig) string {
	return configFieldSet(c).dumpEnv()
}

// GenerateDoc generates a markdown document with a table describing
// all the configuration settings.
func GenerateDoc() string {
	return configFieldSet(&virtlet_v1.VirtletConfig{}).generateDoc()
}

func mappingMatches(cm virtlet_v1.VirtletConfigMapping, nodeName string, nodeLabels map[string]string) bool {
	if cm.Spec.Config == nil {
		return false
	}
	if cm.Spec.NodeName != "" && cm.Spec.NodeName != nodeName {
		return false
	}
	for label, value := range cm.Spec.NodeSelector {
		actual, found := nodeLabels[label]
		if !found || actual != value {
			return false
		}
	}
	return true
}

// MergeConfigs merges several Virtlet configs together, with
// configs going later taking precedence.
func MergeConfigs(configs []*virtlet_v1.VirtletConfig) *virtlet_v1.VirtletConfig {
	var cfg *virtlet_v1.VirtletConfig
	for _, cur := range configs {
		if cfg == nil {
			cfg = cur
		} else {
			Override(cfg, cur)
		}
	}
	return cfg
}

// Binder is used to extract Virtlet config from a FlagSet.
type Binder struct {
	flagSet   *flag.FlagSet
	config    *virtlet_v1.VirtletConfig
	fieldSet  *fieldSet
	lookupEnv envLookup
}

// NewBinder returns a new Binder.
func NewBinder(flagSet *flag.FlagSet) *Binder {
	config := &virtlet_v1.VirtletConfig{}
	fs := configFieldSet(config)
	fs.applyDefaults()
	if flagSet != nil {
		fs.addFlags(flagSet)
	}
	return &Binder{
		flagSet:  flagSet,
		config:   config,
		fieldSet: fs,
	}
}

// GetConfig returns the config that only includes the fields that
// were explicitly set in the flags. It should be called after parsing
// the flags.
func (b *Binder) GetConfig() *virtlet_v1.VirtletConfig {
	b.fieldSet.clearFieldsNotInFlagSet(b.flagSet)
	b.fieldSet.setFromEnv(b.lookupEnv)
	return b.config
}

// configForNode gets virtlet_v1.VirtletConfig for the specified node name and labels.
func configForNode(mappings []virtlet_v1.VirtletConfigMapping, localConfig *virtlet_v1.VirtletConfig, nodeName string, nodeLabels map[string]string) *virtlet_v1.VirtletConfig {
	cfg := GetDefaultConfig()
	var sortedMappings []virtlet_v1.VirtletConfigMapping
	for _, m := range mappings {
		if mappingMatches(m, nodeName, nodeLabels) {
			sortedMappings = append(sortedMappings, m)
		}
	}
	sort.Slice(sortedMappings, func(i, j int) bool {
		a, b := sortedMappings[i], sortedMappings[j]
		// Items that go later in the list take precedence.
		return a.Spec.Priority < b.Spec.Priority
	})

	configs := []*virtlet_v1.VirtletConfig{cfg}
	for _, m := range sortedMappings {
		configs = append(configs, m.Spec.Config)
	}
	if localConfig != nil {
		configs = append(configs, localConfig)
	}
	return MergeConfigs(configs)
}

// NodeConfig is used to retrieve Virtlet configuration for the current
// node.
type NodeConfig struct {
	clientCfg     clientcmd.ClientConfig
	kubeClient    kubernetes.Interface
	virtletClient virtletclient.Interface
}

// NewNodeConfig creates a new NodeConfig
func NewNodeConfig(clientCfg clientcmd.ClientConfig) *NodeConfig {
	return &NodeConfig{clientCfg: clientCfg}
}

func (nc *NodeConfig) setup() error {
	if nc.kubeClient != nil {
		return nil
	}

	config, err := nc.clientCfg.ClientConfig()
	if err != nil {
		return err
	}

	kubeClient, err := kubernetes.NewForConfig(config)
	if err != nil {
		return fmt.Errorf("can't create kubernetes api client: %v", err)
	}

	virtletClient, err := virtletclient.NewForConfig(config)
	if err != nil {
		return fmt.Errorf("can't create Virtlet api client: %v", err)
	}

	nc.kubeClient = kubeClient
	nc.virtletClient = virtletClient
	return nil
}

// LoadConfig loads the configuration for the specified node.
func (nc *NodeConfig) LoadConfig(localConfig *virtlet_v1.VirtletConfig, nodeName string) (*virtlet_v1.VirtletConfig, error) {
	if err := nc.setup(); err != nil {
		return nil, err
	}

	node, err := nc.kubeClient.CoreV1().Nodes().Get(nodeName, meta_v1.GetOptions{})
	if err != nil {
		return nil, fmt.Errorf("can't get node info for node %q: %v", nodeName, err)
	}

	mappingList, err := nc.virtletClient.VirtletV1().VirtletConfigMappings("kube-system").List(meta_v1.ListOptions{})
	if err != nil {
		return nil, fmt.Errorf("failed to list Virtlet config mappings: %v", err)
	}

	return configForNode(mappingList.Items, localConfig, nodeName, node.Labels), nil
}
